backup.sh

```diff
@@ -32,50 +32,77 @@
+    rm -rv /tmp/etcd-backup
+else
+  # set proper umask
+  umask "${ETCD_BACKUP_UMASK}"
+
+  # validate expire type
+  case "${ETCD_BACKUP_EXPIRE_TYPE}" in
+      days|count|never) ;;
+      *) echo "backup.expiretype needs to be one of: days,count,never"; exit 1 ;;
   esac
-fi
 
-# make dirname and cleanup paths
-BACKUP_FOLDER="$( date "${ETCD_BACKUP_DIRNAME}")" || { echo "Invalid backup.dirname" && exit 1; }
-BACKUP_PATH="$( realpath -m "${ETCD_BACKUP_SUBDIR}/${BACKUP_FOLDER}" )"
-BACKUP_PATH_POD="$( realpath -m "/backup/${BACKUP_PATH}" )"
-BACKUP_ROOTPATH="$( realpath -m "/backup/${ETCD_BACKUP_SUBDIR}" )"
-
-# make nescesary directorys
-mkdir -p "/tmp/etcd-backup"
-mkdir -p "${BACKUP_PATH_POD}"
-
-# create backup to temporary location
-ETCDCTL_API=3 etcdctl --endpoints "${ENDPOINT}:2379" --cacert='/etc/kubernetes/pki/etcd-ca/ca.crt'  --cert='/etc/kubernetes/pki/etcd-peer/tls.crt' --key='/etc/kubernetes/pki/etcd-peer/tls.key' snapshot save /tmp/etcd-backup/snapshot.db
-ETCDCTL_API=3 etcdutl --write-out=table snapshot status /tmp/etcd-backup/snapshot.db
-
-# move files to pvc and delete temporary files
-mv /tmp/etcd-backup/* "${BACKUP_PATH_POD}"
-rm -rv /tmp/etcd-backup
-
-# expire backup
-if [ "${ETCD_BACKUP_EXPIRE_TYPE}" = "days" ]; then
-  find "${BACKUP_ROOTPATH}" -mindepth 1 -maxdepth 1  -type d -mtime "+${ETCD_BACKUP_KEEP_DAYS}" -exec rm -rv {} +
-elif [ "${ETCD_BACKUP_EXPIRE_TYPE}" = "count" ]; then
-  # shellcheck disable=SC3040,SC2012
-  ls -1tp "${BACKUP_ROOTPATH}" | awk "NR>${ETCD_BACKUP_KEEP_COUNT}" | xargs -I{} rm -rv "${BACKUP_ROOTPATH}/{}"
+  # validate  expire numbers
+  if [ "${ETCD_BACKUP_EXPIRE_TYPE}" = "days" ]; then
+    case "${ETCD_BACKUP_KEEP_DAYS}" in
+      ''|*[!0-9]*) echo "backup.expiredays needs to be a valid number"; exit 1 ;;
+      *) ;;
+    esac
+  elif [ "${ETCD_BACKUP_EXPIRE_TYPE}" = "count" ]; then
+    case "${ETCD_BACKUP_KEEP_COUNT}" in
+      ''|*[!0-9]*) echo "backup.expirecount needs to be a valid number"; exit 1 ;;
+      *) ;;
+    esac
+  fi
+
+  # make dirname and cleanup paths
+  BACKUP_FOLDER="$( date "${ETCD_BACKUP_DIRNAME}")" || { echo "Invalid backup.dirname" && exit 1; }
+  BACKUP_PATH="$( realpath -m "${ETCD_BACKUP_SUBDIR}/${BACKUP_FOLDER}" )"
+  BACKUP_PATH_POD="$( realpath -m "/backup/${BACKUP_PATH}" )"
+  BACKUP_ROOTPATH="$( realpath -m "/backup/${ETCD_BACKUP_SUBDIR}" )"
+
+  # make nescesary directorys
+  mkdir -p "/tmp/etcd-backup"
+  mkdir -p "${BACKUP_PATH_POD}"
+
+  # create backup to temporary location
+  ETCDCTL_API=3 etcdctl --endpoints "${ENDPOINT}:2379" --cacert='/etc/kubernetes/pki/etcd-ca/ca.crt'  --cert='/etc/kubernetes/pki/etcd-peer/tls.crt' --key='/etc/kubernetes/pki/etcd-peer/tls.key' snapshot save /tmp/etcd-backup/snapshot.db
+  ETCDCTL_API=3 etcdutl --write-out=table snapshot status /tmp/etcd-backup/snapshot.db
+
+  # move files to pvc and delete temporary files
+  mv /tmp/etcd-backup/* "${BACKUP_PATH_POD}"
+  rm -rv /tmp/etcd-backup
+
+  # expire backup
```

Comment:
Do we want to manage the removal of old backups in there too with `mc` maybe? Then we would not need of a retention policy, what do you think?

Reply:
I don't know, I mostly focused on the first if statement, which publishes to s3. This part is responsible for nfs backups. 

This part is edited, because I had to format it and that's all

Reply:
Right, well we could add that later if needed, so for now I suggest we are fine! Let's merge and test this then :)